### PR TITLE
Always load ALL child products for a configurable product

### LINKED ISSUE #25287: LinkManagement::getChildren() returns only children products which are in stock
### Preconditions (*)
1. Version 2.3.x & 2.3-develop
2. Class `Magento\ConfigurableProduct\Mode\LinkManagement`

### Steps to reproduce (*)
1. Inject LinkManagement into any custom code.
2. Call getChildren() with an SKU of a given configurable product (for example let's say it has 5 child simple products, where 3 are in stock and 2 are out of stock.)

```
// ex. product with sku '1234' is a known configurable product with 5 simple product children
$childProducts = $this->linkManagement->getChildren("1234");
echo sizeof($childProducts);
// 3.... (further inspection shows it is the 3 products in stock)
```

### Expected result (*)
The children array should include ALL 5 children of the configurable product, as the method description states in LinkManagementInterface: `Get all children for Configurable product`

### Actual result (*)
The children array only includes the children which are in stock

I looked deeper into the stack, and found in

Magento\ConfigurableProduct\Model\Product\Type\Configurable::getUsedProducts()

which calls

Magento\ConfigurableProduct\Model\Product\Type\Configurable::loadUsedProducts()

which has the $isSaleable parameter. It appears to correctly be called with $isSaleable set to false (i.e., we want all children, not just the ones that are non-salable, which an out of stock product would be) 

In this case it uses the $cacheKey, which then returns the false list.

I did a second sanity check with a freshly cleaned and flushed cache, but the result remained the same.

So my guess is the error lies somewhere in the logic of how $cacheKey is generated - somewhere there is a setting restricting the products.

The only way I've found to get a the resulting true complete set of child products for a configurable product is with another method `getConfiguredUsedProductCollection()` in the same class and the 'skipStockFilter' parameter set to true:

$this->getConfiguredUsedProductCollection($product, true)

Actually this should anyway be the case in loadUsedProducts, when upon an invalidated cache call,

`$usedProducts = $this->readUsedProductsCacheData($cacheKey);`

will return null, and so getConfiguredUsedProductCollection should be called with the true option. Even when saleableOnly is set to true this would be okay, because there is the further step of processing the return collection (which I am assuming correctly cuts out non-saleable products)

```
if ($salableOnly) {
     $collection = $this->salableProcessor->process($collection);
}
```

I'd be happy to do a pull request on this if my logic is sound agreed upon.

### PR DESCRIPTION
**Pull request by princefishthrower**
Always load ALL child products for a configurable product (set `skipStockFilter` to 'true')

### Description (*)
This small pull request only changes a passed parameter from false to true. I discovered that the LinkManagement class was returning only children of configurable products that were 'in stock' for a given configurable product. However, the behaviour described for the getChildren() method states that all children should be returned. 

I discovered that the `skipStockFilter` flag was falsely set to `false`, which causing this behavior.

### Fixed Issues
1. Fixes magento/magento2#25287 : LinkManagement::getChildren() returns only children products which are in stock

https://github.com/magento/magento2/issues/25287

### Manual testing scenarios (*)
1. Create a configurable product (for illustration purposes, lets say with SKU '1234') with multiple simple product children
2. Set any number (or even all!) children to 'out of stock'
3. Injecting LinkManagement in any custom code and calling `$this->linkManagement->getChildren('1234')` will now return ALL child products - and subsequent cached calls likewise.

Previous behavior: `$this->linkManagement->getChildren('1234')` returned only in stock children (and subsequent cached calls likewise).

### Contribution checklist (*)
 - [X] Pull request has a meaningful description of its purpose
 - [X] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
Hi @princefishthrower , please, sign CLA, otherwise, we can't process your pull request.
Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. Please cover your fix by automated tests.

**Comment by princefishthrower** [comment]
@VladimirZaets - thanks, I've signed the CLA.

For the tests, is there a link or article as to which proper automated tests I should run and how to run them?

**Comment by VladimirZaets** [comment]
@princefishthrower we run tests automatically, you can check results of the tests in GitHub checks.
In this case, you change the logic, so you should cover this logic with any type of automated tests (unit/functional/integration). You can find guidelines and examples how to write new tests on http://devdocs.magento.com/

**Comment by engcom-Echo** [comment]
I will take care of test coverage.

**Comment by engcom-Echo** [comment]
Failure of functional test, not related to the changes in this PR , test failing on 2.4-develop branch
image

File: app/code/Magento/ConfigurableProduct/Test/Unit/Model/Product/Type/ConfigurableTest.php [hunk 1]
```diff
@@ -230,17 +227,101 @@ protected function setUp()
...
+    {
+        $storeId = 1;
+        $customerGroupId = 1;
+        $productId = 1;
+        $linkField = 'id';
+        $product = $this->createMock(Product::class);
+        $productCollectionItem = $this->createMock(Product::class);
+        $productCollection = $this->createMock(ProductCollection::class);
+        $this->entityMetadata->expects($this->atLeastOnce())
+        ->method('getLinkField')
```
Comment (Nazar65) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Model/Product/Type/ConfigurableTest.php]:
```suggestion
            ->method('getLinkField')
```

File: app/code/Magento/ConfigurableProduct/Test/Unit/Model/Product/Type/ConfigurableTest.php [hunk 2]
```diff
@@ -230,17 +227,101 @@ protected function setUp()
...
 
-    public function testHasWeightTrue()
+    /**
+     * Verify load user product if stock filter is true.
+     *
+     * @return void
+     */
+    public function testLoadUserProductsSkipStockFilterIsTrue(): void
+    {
+        $storeId = 1;
```
Comment (Nazar65) [inline app/code/Magento/ConfigurableProduct/Test/Unit/Model/Product/Type/ConfigurableTest.php]: Could you please move data to dataProvider ?

**Review by Nazar65 — CHANGES_REQUESTED** [review]

**Comment by engcom-Echo** [comment]
Hi @VladimirZaets could you please approve it?

**Comment by engcom-Echo** [comment]
Hi @VladimirZaets could you please approve it?

**Comment by engcom-Echo** [comment]
Hi @VladimirZaets could you please approve it?

**Comment by VladimirZaets** [comment]
@engcom-Echo can you please resolve conflicts?

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
Hi @VladimirZaets could you please approve it?

**Review by ihor-sviziev — DISMISSED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8086](https://jira.corp.magento.com/browse/ENGCOM-8086) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Review by engcom-Alfa — CHANGES_REQUESTED** [review]
Hi @princefishthrower 
Unfortunately, we are not able to reproduce this [issue](https://github.com/magento/magento2/issues/25287) on a fresh 2.4-develop

**Precondition:**

Configurable product with 5 children ( 3 -  In Stock, 2 - Out of Stock)

**Manual testing scenario:**

1. Inject LinkManagement into any custom code (Class `Magento\ConfigurableProduct\Mode\LinkManagement`)
2. Call getChildren() with an SKU of a given configurable product (for example let's say it has 5 child simple products, where 3 are in stock and 2 are out of stock.)

```
// ex. product with sku 'configurable2' is a known configurable product with 5 simple product children
$childProducts = $this->linkManagement->getChildren("configurable2");
echo sizeof($childProducts);
// 5.... (further inspection shows all 5 products)
```

**Actual Result:** :heavy_check_mark: 5. The children array include ALL 5 children of the configurable product
Screenshot from 2020-09-03 14-58-21

@princefishthrower Could you take a look?

Thanks!

**Comment by gabrieldagama** [comment]
The risk was set to`low` due to: the suggested changes shouldn't affect other areas.

**Comment by ihor-sviziev** [comment]
@princefishthrower could you confirm that this issue isn't reproducible anymore or add some additional info in order to reproduce it?
Thank you!

**Comment by sidolov** [comment]
@princefishthrower , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for the collaboration!

**Pull request closed without merging**